examples/twittermap/web/public/javascripts/heatmapcache/services.js

```diff
@@ -0,0 +1,50 @@
+'use strict';
+angular.module('cloudberry.heatmapcache', ['cloudberry.common'])
+  .service('heatMapResultCache', function () {
+
+  // The key-value store that stores heatmap results of a query.
+  var store = [];
+  // To check if time range in query changed (new query?)
+  var endDate = new Date();
+  // Deducts -1 to create a default value that can't exist
+  endDate.setDate(endDate.getDate() - 1);
+  var currentTimeRange = {
+    start: new Date(),
+    end: endDate
+  };
+
+  this.emptyStore = function () {
+    store = [];
+    currentTimeRange.start = new Date();
+    currentTimeRange.end = new Date();
+  }
+
+  this.cacheIsDone = function(timeInterval) {
+    return currentTimeRange.end >= timeInterval.end && currentTimeRange.start <= timeInterval.start;
+  }
+
+  /**
+   * Retrieves map results data from the cache; ignores empty objects
+  */
+  this.getValues = function (timeInterval) {
+    var resultArray = [];
+    for (var j = 0; j < store.length; j++) {
+      var day = new Date(store[j]["create_at"]);
+      if (day >= timeInterval.start && day <= timeInterval.end) {
+        resultArray.push(store[j]);
+      }
+    }
+    return resultArray;
+  };
+
+  /**
+   * Updates the store with map result each time the middleware responds to json request.
```

Comment:
If each time gets a partial result, we `putValues` the results into cache, the logic of updating `currentTimeRange` is not correct. It should be extended instead of assigned the `timeInterval` of current partial results.

Reply:
I have changed it to the partial time range, but I realized there is a small problem with it. For example, since keyword 'Trump' has a large amount of data, it will only fetch data for one to two months. In this case, it will never be considered as 'complete heatmap cache hit case' because the entire time range is one year, but the cache time range is only one or two months.

Reply:
Sorry, I didn't follow. Can you elaborate more?